lib/streamlit/runtime/scriptrunner/script_runner.py

```diff
@@ -593,13 +588,14 @@ def _run_script(self, rerun_data: RerunData) -> None:
             self._run_script(rerun_exception_data)
 
     def _on_script_finished(
-        self, ctx: ScriptRunContext, event: ScriptRunnerEvent
+        self, ctx: ScriptRunContext, event: ScriptRunnerEvent, premature_stop: bool
     ) -> None:
         """Called when our script finishes executing, even if it finished
         early with an exception. We perform post-run cleanup here.
         """
         # Tell session_state to update itself in response
-        self._session_state.on_script_finished(ctx.widget_ids_this_run)
+        if not premature_stop:
```

Comment:
If we are handling this for StopException/Rerun Exception to indicate the use of st.stop/rerun, don't we want to update the session state and handle it?

I may be perhaps misunderstanding the value of ignoring the change.

Reply:
Currently, if a stop/rerun comes in and fast reruns is enabled (which is ~always the case, very very few apps turn it off), `on_script_finished` will be a no-op for the disconnected session state. This prevents widgets that were going to be rendered, but didn't because it was interrupted, from losing their state. To retain this behavior, since it no longer happens via the disconnect mechanism, we have to conditionally avoid the session state cleanup step via the script runner instead.

Reply:
Let me walk through an example and make sure I understand.

Let's look at the following script:

```python
import streamlit as st

st.button("misdirection")

if st.button("belly"):
    st.write("Here's an issue")
    st.stop()

st.write("This may not get written if the button is pressed")
```

When we run it, no major issue. When we click the button, we do the following:
1. The button's value should be set to true (trigger)
1. The script executes. The button flow runs and `st.stop()` is called.
1. We [request to stop](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/commands/execution_control.py#L46)
1. (This part may be a change from my original understanding) That eventually raises an Exception, specifically a StopException, right?
1. The script finishes and calls on_script_finished.
1. It should call session state on script finished because we want to [reset the button trigger](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/state/session_state.py#L537)? Or to remove the stale widgets (the example doesn't provide this scenario)?

Reply:
1-4 is still correct.
5. For the code on `develop`, if fast reruns is enabled, `on_script_finished` runs but does nothing (immediate return if disconnected). For the PR code, `on_script_finished` is instead not called at all.
6. The button trigger gets reset at the beginning of the next script run, due to a change I made recently for a bug fix. Possibly we should remove the trigger reset call from `on_script_finished`, to simplify things.

Reply:
Okay. This makes more sense now. Would love if we can try to simplify the flow for understanding.